crates/data_components/src/iceberg/catalog/mod.rs

```diff
@@ -42,7 +42,7 @@ pub enum Error {
     NamespaceDoesNotExist { namespace: String },
 
     #[snafu(display(
-        "Failed to connect to the Iceberg catalog or object store at {url}, verify the Iceberg catalog is accessible and try again."
+        "Failed to connect to the Iceberg catalog or object store at {url}, verify the Iceberg catalog is accessible and try again. Connection failure caused by: {source:?}"
```

Comment:
```suggestion
        "Failed to connect to the Iceberg catalog or object store at {url}: {source}. Verify the Iceberg catalog is accessible and try again."
```
Our [error handling guidelines](https://github.com/spiceai/spiceai/blob/trunk/docs/dev/error_handling.md) say that we should avoid including debug information in our error messages. Could you see if the error message I suggested above surfaces the error correctly?

Reply:
@Advayp 
Unfortunately the error in this case is still a bit vague and doesn't point to the certificate issues.

```rust
2025-08-25T19:39:22.223020Z DEBUG hyper_util::client::legacy::connect::http: connecting to x.x.x.x:443
2025-08-25T19:39:22.515759Z DEBUG hyper_util::client::legacy::connect::http: connected to x.x.x.x:443
2025-08-25T19:39:23.096868Z TRACE hyper_util::client::legacy::pool: checkout dropped for ("https", catalog.yourdomain.com/catalog/v1/config)
2025-08-25T19:39:23.097440Z ERROR runtime::init::catalog: Failed to setup the catalog lakekeeper (iceberg).
Failed to connect to the Iceberg catalog or object store at (https://catalog.yourdomain.com/catalog/v1/config): Unexpected => Failed to execute http request, source: error sending request for url (https://catalog.yourdomain.com/catalog/v1/config). Verify the Iceberg catalog is accessible and try again.
2025-08-25T19:40:20.419494Z  INFO runtime::init::catalog: Registering catalog 'lakekeeper' for iceberg
```

The extra information as seen below when using `{source:?}` is **missing** in the error log now.
```rust
Source: error sending request for url (https://catalog.yourdomain.com/catalog/v1/config): client error (Connect): A certificate chain processed, but terminated in a root certificate which is not trusted by the trust provider. (os error -2146762487)
```
How would you suggest to proceed on this? It would be pretty useful knowing that the error is being caused by cert issues.

Reply:
Thanks for checking that @ChrisTomAlxHitachi. @peasee and I think that it would be appropriate to handle the specific underlying variant directly. The lack of non-debug information means that the iceberg's crate [error struct's](https://docs.rs/iceberg-rust/latest/iceberg_rust/error/enum.Error.html) variants aren't specific enough, so we should match for the specific underlying variant that causes the certificate error, map it to a new variant in the `Error` struct in `crates/data_components/src/iceberg/catalog/mod.rs`, and return that instead.

Reply:
@Advayp @peasee 
Updated the PR with a new Snafu. Is this better? Anything I can do to improve it?

Reply:
Thanks for re-working it. It isn't exactly what I had in mind, as we're still displaying the debug error message from `format!("{e:?}")` into `CertificateError { url: String, detail: String, source: iceberg::Error }`. I think we can get more specific if we attempt downcasting the iceberg error source into the specific reqwest error that might be throwing this. I had an attempt at a patch:

```patch
 crates/data_components/src/iceberg/provider.rs | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/crates/data_components/src/iceberg/provider.rs b/crates/data_components/src/iceberg/provider.rs
index 75306648..c5a19a24 100644
--- a/crates/data_components/src/iceberg/provider.rs
+++ b/crates/data_components/src/iceberg/provider.rs
@@ -18,6 +18,7 @@ limitations under the License.
 
 use std::any::Any;
 use std::collections::HashMap;
+use std::error::Error as StdError;
 use std::sync::Arc;
 
 use async_trait::async_trait;
@@ -274,6 +275,18 @@ fn handle_iceberg_error(e: iceberg::Error) -> Error {
         iceberg::ErrorKind::Unexpected => {
             // This is also returned when we cannot connect to the Iceberg catalog, so check for that.
             // i.e. Unexpected => Failed to execute http request, source: error sending request for url (http://localhoster:8181/v1/config)
+
+            if let Some(source) = e.source() {
+                if let Some(reqwest_err) = source.downcast_ref::<reqwest::Error>() {
+                    if reqwest_err.is_request() {
+                        if let Some(source) = reqwest_err.source() {
+                           let inner_msg = format!("{source:?}");
+                           if inner_msg.contains("not trusted") && inner_msg.contains("trust provider") {
+                               // UntrustedCertificateSnafu
+                           }
+                        }
+                    }
+                }
+            }
+
             let err_msg = e.to_string();
             let err_in_detail = format!("{e:?}");
             let err_in_detail_lc = err_in_detail.to_lowercase();
```

If we combine this with a new error variant like `UntrustedCertificate { url: String }` I think we can get a better result with a more descriptive error. This is primarily because `reqwest::Error` does not implement a `Display` utilising the underlying error: https://github.com/seanmonstar/reqwest/blob/master/src/error.rs#L226

Let me know what you think.

Reply:
@peasee  
That makes sense. Thank you for the code block you shared, let me try to do the same on my end.

Reply:
@peasee 
Were you able to get this patch running with proper error downcasting? I looped through the error chain and still the downcast never worked. I suspect it might be some issue on my end if you successfully downcasted the error, I will give it another try soon.

Reply:
Yes, I must admit I never attempted it because I did not have a reproducible example. If downcasting does not work, then let's stick with your original example for now, but instead of a generic `CertificateError` can we string match on the debug error for the specific untrusted certificate error? This would let us avoid displaying the debug error in the output.